### ISSUE TITLE
[A11Y] Cannot manually set focus after opening amp-sidebar

### ISSUE THREAD

**Issue by jamesshannon** [ampproject/amphtml#12240]
AMP supports the focus action, which allows for the setting of focus. This is Good for accessibility.

I've got a partner who's very keen on A11Y and has the following requirement:

> After amp-sidebar is shown focus needs to be on a specific element within the sidebar, and not the sidebar itself.

Currently, opening amp-sidebar sets focus on the sidebar itself. Surprisingly, HTML doesn't support a "pass focus to first tabindex child" attribute. So my thought was to use the focus action to manually set focus. While that action works in isolation, it doesn't have any effect after amp-sidebar is opened. It should set focus as expected.

Test Page: http://jsbin.com/mifesek/2/edit?html,output

1. Click the 'Focus Test' button. See that the X <amp-img> is highlighted in yellow (via CSS :focus rule). This is WAI.

2. Click the 'Toggle Sidebar' button. This is a standard sidebar example from ampbyexample. The sidebar gets focus (turns yellow), which is exactly what happens on ampbyexample.

I've added tabindex=-1 to amp-sidebar in hopes that this would prevent focus. It has no effect.

Additionally, the same button that toggles the sidebar also tries to set focus on the close-btn.. This also has no effect.

**Comment by aghassemi** [comment]
@jamesshannon Normally the best practice is to focus on the sidebar itself and communicate via `title` a summary of the sidebar. I can see however that in some cases it is more desirable to focus on a particular element in the sidebar.

There is a race-condition bug here with calling open then focus. Will look into it.

**Comment by jamesshannon** [comment]
Partner's requirement (for whatever reason) is to focus on the close button. Personally, I think that the sidebar focus makes more sense. But getting .focus working would allow the behavior to be manually set in code.

I think the biggest issue is the iOS bug in which the focus stays on the clicked element. That could also be worked around if .focus worked.

**Comment by jamesshannon** [comment]
@aghassemi Have you had a chance to look at the race condition?

**Comment by aghassemi** [comment]
Hi @jamesshannon , we will be picking this issue up this Friday in our a11y hackathon day. I am very hopeful a fix will make it to next week's Dev Channel release on Tuesday.

**Comment by aghassemi** [comment]
@jamesshannon We endedup adding a `sidebarOpen` event that can be used to focus on an element:
```
<amp-sidebar id="sidebar1" layout="nodisplay" on="sidebarOpen:focusOnMe.focus">
```

This new approach was needed because browser also refuse to focus on invisible element, so we need to make sure sidebar is fully visible before focusing.

This change will make it to Dev Channel tomorrow or Wednesday but due to holiday freeze may not make it to production until January 4th.
